packages/@lwc/engine-core/src/framework/modules/static-parts.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { isNull, isUndefined, assert } from '@lwc/shared';
+import { VStatic, VStaticPart } from '../vnodes';
+import { RendererAPI } from '../renderer';
+import { applyEventListeners } from './events';
+import { applyRefs } from './refs';
+
+function traverseAndSetElements(root: Element, parts: VStaticPart[], renderer: RendererAPI): void {
+    const numParts = parts.length;
+
+    // Optimization given that, in most cases, there will be one part, and it's just the root
+    if (numParts === 1) {
+        const firstPart = parts[0];
+        if (firstPart.partId === 0) {
+            // 0 means the root node
+            firstPart.elm = root;
+            return;
+        }
+    }
+
+    const partIdsToParts = new Map<number, VStaticPart>();
+    for (const staticPart of parts) {
+        partIdsToParts.set(staticPart.partId, staticPart);
+    }
+
+    let numFoundParts = 0;
+    const { previousSibling, getLastChild } = renderer;
+    const stack = [root];
+    let partId = -1;
+
+    // Depth-first traversal. We assign a partId to each element, which is an integer based on traversal order.
+    while (stack.length > 0) {
+        const elm = stack.shift()!;
+        partId++;
+
+        const part = partIdsToParts.get(partId);
+        if (!isUndefined(part)) {
+            part.elm = elm;
+            if (++numFoundParts === numParts) {
+                return; // perf optimization - stop traversing once we've found everything we need
+            }
+        }
+
+        // For depth-first traversal, prepend to the stack in reverse order
+        // Note that we traverse using `*Child`/`*Sibling` rather than `children` because the browser uses a linked
+        // list under the hood to represent the DOM tree, so it's faster to do this than to create an underlying array
+        // by calling `children`.
+        let child = getLastChild(elm);
+        while (!isNull(child)) {
+            stack.unshift(child);
+            child = previousSibling(child);
+        }
```

Comment:
I could have avoided having to add `previousSibling` to the `Renderer` by using `firstChild`/`nextSibling` instead, but then I would have had to either create a temporary array and do `stack.unshift(...tempArray)`or use `stack.splice()`. Neither option seemed great perf-wise. In any case this function takes up like 1ms total in our entire Karma benchmark.

Reply:
TODO: benchmark which is fastest:

1. previousSibling/lastChild/unshift
2. nextSibling/firstChild/unshift from temporary array
3. nextSibling/firstChild/splice